// @nuxtjs/supabase/dist/module.mjs
import { fileURLToPath } from 'node:url';
import fs from 'node:fs';
import { defu } from 'defu';
import { relative } from 'pathe';
import { defineNuxtModule, useLogger, createResolver, addPlugin, addImportsDir, addTemplate, extendViteConfig } from '@nuxt/kit';

const module$1 = defineNuxtModule({
  meta: {
    name: "@nuxtjs/supabase",
    configKey: "supabase",
    compatibility: {
      nuxt: ">=3.0.0"
    }
  },
  defaults: {
    // Env var resolution order: NUXT_PUBLIC_*/NUXT_* (recommended) → SUPABASE_* (legacy fallback) → undefined
    url: process.env.NUXT_PUBLIC_SUPABASE_URL || process.env.SUPABASE_URL || void 0,
    key: process.env.NUXT_PUBLIC_SUPABASE_KEY || process.env.SUPABASE_KEY || process.env.SUPABASE_PUBLISHABLE_KEY || process.env.SUPABASE_ANON_KEY || process.env.NUXT_PUBLIC_SUPABASE_PUBLISHABLE_KEY || process.env.NUXT_PUBLIC_SUPABASE_ANON_KEY || void 0,
    serviceKey: process.env.NUXT_SUPABASE_SERVICE_KEY || process.env.SUPABASE_SERVICE_KEY || void 0,
    secretKey: process.env.NUXT_SUPABASE_SECRET_KEY || process.env.SUPABASE_SECRET_KEY || process.env.SUPABASE_SERVICE_ROLE_KEY || void 0,
    redirect: true,
    redirectOptions: {
      login: "/login",
      callback: "/confirm",
      exclude: [],
      cookieRedirect: false,
      saveRedirectToCookie: false
    },
    cookieName: "sb",
    cookiePrefix: void 0,
    useSsrCookies: true,
    cookieOptions: {
      maxAge: 60 * 60 * 8,
      sameSite: "lax",
      secure: true
    },
    types: "~/types/database.types.ts",
    clientOptions: {}
  },
  setup(options, nuxt) {
    const logger = useLogger("@nuxt/supabase");
    const { resolve, resolvePath } = createResolver(import.meta.url);
    nuxt.options.runtimeConfig.public.supabase = defu(nuxt.options.runtimeConfig.public.supabase, {
      url: options.url,
      key: options.key,
      redirect: options.redirect,
      redirectOptions: options.redirectOptions,
      cookieName: options.cookieName,
      cookiePrefix: options.cookiePrefix,
      useSsrCookies: options.useSsrCookies,
      cookieOptions: options.cookieOptions,
      clientOptions: options.clientOptions
    });
    nuxt.options.runtimeConfig.supabase = defu(nuxt.options.runtimeConfig.supabase || {}, {
      serviceKey: options.serviceKey,
      secretKey: options.secretKey
    });
    const finalUrl = nuxt.options.runtimeConfig.public.supabase.url;
    if (!finalUrl) {
      logger.warn("Missing `NUXT_PUBLIC_SUPABASE_URL`, set it in `.env` or via `runtimeConfig.public.supabase.url` in `nuxt.config.ts`.");
    } else {
      try {
        const defaultStorageKey = `sb-${new URL(finalUrl).hostname.split(".")[0]}-auth-token`;
        const currentPrefix = nuxt.options.runtimeConfig.public.supabase.cookiePrefix;
        nuxt.options.runtimeConfig.public.supabase.cookiePrefix = currentPrefix || defaultStorageKey;
      } catch (error) {
        logger.error(
          `Invalid Supabase URL: "${finalUrl}". Please provide a valid URL (e.g., https://example.supabase.co or http://localhost:5432)`,
          error
        );
        const currentPrefix = nuxt.options.runtimeConfig.public.supabase.cookiePrefix;
        nuxt.options.runtimeConfig.public.supabase.cookiePrefix = currentPrefix || "sb-auth-token";
        if (!nuxt.options.dev) {
          throw new Error(`Invalid Supabase URL: "${finalUrl}". Provide a valid URL or leave it unset for runtime configuration.`);
        }
      }
    }
    if (!nuxt.options.runtimeConfig.public.supabase.key) {
      logger.warn("Missing `NUXT_PUBLIC_SUPABASE_KEY`, set it in `.env` or via `runtimeConfig.public.supabase.key` in `nuxt.config.ts`.");
    }
    if (nuxt.options.runtimeConfig.public.supabase.redirectOptions.cookieRedirect) {
      logger.warn("The `cookieRedirect` option is deprecated, use `saveRedirectToCookie` instead.");
    }
    if (nuxt.options.runtimeConfig.public.supabase.cookieName != "sb") {
      logger.warn("The `cookieName` option is deprecated, use `cookiePrefix` instead.");
    }
    const supabaseConfig = nuxt.options.runtimeConfig.supabase;
    const hasServiceKey = !!supabaseConfig?.serviceKey;
    const hasSecretKey = !!supabaseConfig?.secretKey;
    if (hasServiceKey && !hasSecretKey) {
      logger.warn("`SUPABASE_SERVICE_KEY` is deprecated. Migrate to `NUXT_SUPABASE_SECRET_KEY`. See: https://supabase.com/blog/jwt-signing-keys");
    }
    const mergedOptions = nuxt.options.runtimeConfig.public.supabase;
    if (mergedOptions.redirect && mergedOptions.redirectOptions.callback) {
      const routeRules = {};
      routeRules[mergedOptions.redirectOptions.callback] = { ssr: false };
      nuxt.options.nitro = defu(nuxt.options.nitro, {
        routeRules
      });
    }
    const runtimeDir = fileURLToPath(new URL("./runtime", import.meta.url));
    nuxt.options.build.transpile.push(runtimeDir);
    addPlugin(resolve(runtimeDir, "plugins", "supabase.client"));
    addPlugin(resolve(runtimeDir, "plugins", "supabase.server"));
    if (mergedOptions.redirect) {
      addPlugin(resolve(runtimeDir, "plugins", "auth-redirect"));
    }
    addImportsDir(resolve("./runtime/composables"));
    nuxt.hook("nitro:config", (nitroConfig) => {
      nitroConfig.alias = nitroConfig.alias || {};
      nitroConfig.externals = defu(typeof nitroConfig.externals === "object" ? nitroConfig.externals : {}, {
        inline: [resolve("./runtime")]
      });
      nitroConfig.alias["#supabase/server"] = resolve("./runtime/server/services");
      nitroConfig.alias["#supabase/database"] = resolve(nitroConfig.buildDir, "types/supabase-database");
    });
    addTemplate({
      filename: "types/supabase.d.ts",
      getContents: () => [
        "declare module '#supabase/server' {",
        `  const serverSupabaseClient: typeof import('${resolve("./runtime/server/services")}').serverSupabaseClient`,
        `  const serverSupabaseServiceRole: typeof import('${resolve(
          "./runtime/server/services"
        )}').serverSupabaseServiceRole`,
        `  const serverSupabaseUser: typeof import('${resolve("./runtime/server/services")}').serverSupabaseUser`,
        `  const serverSupabaseSession: typeof import('${resolve("./runtime/server/services")}').serverSupabaseSession`,
        "}"
      ].join("\n")
    });
    addTemplate({
      filename: "types/supabase-database.d.ts",
      getContents: async () => {
        if (options.types) {
          try {
            const path = await resolvePath(options.types);
            const typesPath = await resolvePath("~~/.nuxt/types/");
            if (fs.existsSync(path)) {
              return `export * from '${relative(typesPath, path)}'`;
            } else {
              logger.warn(
                `Database types configured at "${options.types}" but file not found at "${path}". Using "Database = unknown".`
              );
            }
          } catch (error) {
            logger.error(`Failed to load Supabase database types from "${options.types}":`, error);
          }
        }
        return `export type Database = unknown`;
      }
    });
    nuxt.hook("prepare:types", async (options2) => {
      options2.references.push({ path: resolve(nuxt.options.buildDir, "types/supabase.d.ts") });
    });
    if (!nuxt.options.dev && !["cloudflare"].includes(process.env.NITRO_PRESET)) {
      nuxt.options.build.transpile.push("websocket");
    }
    extendViteConfig((config) => {
      config.optimizeDeps = config.optimizeDeps || {};
      config.optimizeDeps.include = config.optimizeDeps.include || [];
      config.optimizeDeps.include.push("@nuxtjs/supabase > @supabase/ssr > cookie");
    });
  }
});

export { module$1 as default };
